### PR TITLE
Update build infrastructure

### PR DESCRIPTION
**Pull request by 00-matt**
The old Makefile worked okay, but this one allows for parallel build jobs (make -jN) and includes header files in the dependency graph.

### CODE REVIEW

**Comment by WardCunningham** [comment]
Ouch.

I enjoyed reading the original makefile and learned something doing so. Not so much this one.

**Comment by f-prime** [comment]
@00-matt This is great. The only thing that I ask is that the binary be outputted to the bin/ folder in the parent directory like it is now. I can handle this if needed, let me know. Otherwise this looks and works great, thank you.

**Comment by 00-matt** [comment]
On Saturday, 22 June 2019 17:24:34 BST Ward Cunningham wrote:
> I enjoyed reading the original makefile and learned something doing so. Not
> so much this one.

Do you have any suggestions? It's quite a bit longer than the original 
Makefiles, but I don't think that any of it is scary or hard to follow.

On Sunday, 23 June 2019 02:35:28 BST Frankie Primerano wrote:
> @00-matt This is great. The only thing that I ask is that the binary be
> outputted to the local bin/ directory like it is now. I can handle this if
> needed, let me know. Otherwise this looks and works great, thank you.

Whoops sorry, I should have done that to begin with. Fixed and force pushed 
now :)

**Comment by f-prime** [comment]
Perfect, thank you again for this. Merging into master :)
